### ISSUE TITLE
Support expanded configuration keys

### ISSUE THREAD

**Issue by TDabasinskas** [wintoncode/Winton.Extensions.Configuration.Consul#36]
Hello,

Are there any plans on implementing "Add support for expanded configuration where the configuration is a tree of KV pairs under the root key", mentioned on in the backlog section?

**Comment by Choc13** [comment]
Hi @TDabasinskas,

We did have #7 open for that work, but we found we didn't actually need that behaviour in practice, so I closed the ticket as it had been sat idle for a long time. However, if this is going to be useful for you and you have an idea how it would be implemented, please feel free to submit a PR for it.

**Comment by TDabasinskas** [comment]
Hi, @Choc13,

Yes, I saw #7, but since it was closed and had no additional comments, I raised a new one.

Having support for expanded keys, would allow using [bulkload functionality of Consul-CLI](https://github.com/mantl/consul-cli/wiki/KV#bulkload). In addition, it would allow to maintain the configs more easily, instead of having a one large key in Consul.

I'm afraid, though. I will not be able to submit a PR for this, as it seem to require more than by current skills.

**Comment by Choc13** [comment]
Yeah, I agree with rationale behind adding the feature. I might see if I can put something together, but I can't promise if and when it will be ready. I'll try and do an initial investigation over the next week of so to see if I can get a better gauge on how much work is involved and then let you know from there.

**Comment by Choc13** [comment]
Hi @TDabasinskas,

I've had a bit more of a look into this, and, unless I'm missing a trick this doesn't seem to be so straightforward. For starters, I think the only way to achieve this would be to use the `recurse` query option, which returns a list of all KV nodes in the hierarchy. This in itself is not a problem, but will require some refactoring around how the values are parsed into an appropriate dictionary structure. For instance, currently I would need to write something that adheres to the [`IConfigurationParser`](https://github.com/wintoncode/Winton.Extensions.Configuration.Consul/blob/master/src/Winton.Extensions.Configuration.Consul/Parsers/IConfigurationParser.cs), but that is expecting a `Stream` and not a list of KV pairs.

One option could be to use `recurse` in conjunction with the current parsing infrastructure. What I mean by this, is that it would be possible to create configuration out of keys under a specific root key, but whose values might be encoded as either, a single value, JSON or YAML etc. So for instance we might have the following KV setup:

```
MyApp
    - ConfigSection1
        - AKey : "hello"
    - ConfigSection2 : { "foo": ["bar", "baz"] }
```

Which would be converted to the following in the `IConfiguration` space:

```
MyApp:ConfigSection1:AKey:hello
MyApp:ConfigSection2:foo:0:bar
MyApp:ConfigSection2:foo:1:baz
```

Does this sound correct to you and does it meet your requirements?

One further question I have it around how you would expect arrays to work in a purely single value setup. I'm imaging it that it would be handled much like Microsoft have done in their configuration library. So for example, the following KV setup in Consul

```
MyApp
    - Array
        - 0 : "a"
        - 1 : "b"
```

Would be converted to:

```
MyApp:Array:0:a
MyApp:Array:1:b
```

I can't think of any other way to represent arrays as keys in Consul.

Let me know if the above sounds like the behaviour you were expecting.

**Comment by TDabasinskas** [comment]
Hi @Choc13,

Thanks for finding time to look into this.

The first example of the way config sections and KV would be handled does indeed seem to meet my requirements.

Regarding the arrays, I'm not sure I've understood the following example:

```yaml
MyApp
    - Array
        - 0 : "a"
        - 1 : "b"
```

To my understanding, the array should be considered a value, so in Consul, I would define it as this:

- Directory: `MyApp`
- Key `Array`
- Value `{ ["a", "b"] }`

Or, do you think (based on the example you provided), each array element should be treated as a separate key?

Once parsed, yes, I think they should look like this: `MyApp:Array:0:a`. I assume I would be able to retrieve value of `MyApp:Array` as `string[]` and loop though it (I guess Microsoft options library will handle that).

**Comment by Choc13** [comment]
OK great. I think what I was proposing above would actually allow you to either encode arrays as a JSON formatted value or encode them as 0 indexed integer keys. It would be useful to know how the [bulkload functionality of Consul-CLI](https://github.com/mantl/consul-cli/wiki/KV#bulkload) (that you mentioned above) handles arrays. Do you know? I have never used that tool before and there doesn't appear to be clear instructions on how to install it.

Also, do you ever imagine a scenario where you might want to encode the values in a different format (say YAML)?

**Comment by TDabasinskas** [comment]
Let's assume `C:\Temp\dummy.json` contains the following:

```json
{
	"SomeKey": "SomeValue",
	"SomeHash": {
	  "SomeHashKey": "SomeHashValue",
	  "SomeHashKey2": "SomeHashValue2"
	},
	"SomeArray": [
	  "SomeArrayElement1",
	  "SomeArrayElement2",
	  "SomeArrayElement3"
	]
}
```

We run the following `consul-cli` command:

```
consul-cli kv bulkload --consul=FQDN:443 --ssl=true --datacenter=DC --token=TOKEN--json="C:\Temp\dummy.json" --prefix=MyApp
```

It creates the following keys (a new key is created for each hash key OR array element):

- /MyApp/SomeHash/SomeHashKey
- /MyApp/SomeHash/SomeHashKey2
- /MyApp/SomeKey
- /MyApp/SomeArray/0
- /MyApp/SomeArray/1
- /MyApp/SomeArray/2

As for YAML, I've actually considered using it, but I've noticed this library supports JSON-only, so I've stick to it. It would definitely be nice to have support for YAML, but that's nothing critical (at least for me).

**Comment by Choc13** [comment]
Ok. So it looks like it does encode JSON arrays using zero-indexes integer keys in Consul’s KV store. 

I think I have enough information to start implementing this now. I think we can actually always recurse keys and then auto detect which parser to use for each value in that tree (e.g. text, JSON or YAML). The most obvious risk is that the auto detection is fiddly and doesn’t consist of mutually exclusive rules. 

This strategy should also ensure backwards compatibility with clients just storing JSON config under a single key too, as that is just a special case of what is described above. 

Can’t really give an estimate of when this will be available at this time though as it feels like there are still a few places that this plan might get stuck on.

**Comment by Choc13** [comment]
@TDabasinskas the above PR is the first and I think major step towards supporting this feature. Once that is merged the library will support specifying just a root key and then all KV pairs under that key will be loaded and merged into a single config dictionary. However, to keep things simple and to maintain backwards compatibility I have opted to enforce a restriction that all values be formatted using the same scheme. For now that is JSON, as the library already has a parser for this, but I intend to do a subsequent PR to add a single value parser. Once that is in place then you should have all the pieces necessary to support reading config that you have bulk loaded using the Consul CLI. Again, for backwards compatibility reasons the JSON parser will remain the default, but you would just need to select the SingleValueParser in the options.

**Comment by Choc13** [comment]
Hi @TDabasinskas, 

I got around to implementing `SimpleConfigurationParser` yesterday and it was merged today. Would you be able to try using the latest master build, it is on [NuGet](https://www.nuget.org/packages/Winton.Extensions.Configuration.Consul/2.1.0-master0003) to match it matches your requirements? You will just need to set the parser to `SimpleConfigurationParser` which can be done via the options, e.g.

```
.AddConsul(
    "path/To/Root/Key",
    cancellationTokenSource.Token,
    options =>
    {
        options.Parser = new SimpleConfigurationParser();
        // other options
    })
```

If that works then I will release version 2.1.0 with those changes.

Cheers.
